### ISSUE TITLE
Make in-VM reboot request actually restart the VM

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#8196]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)

R4.2

### The solution you'd like

Make executing reboot from within a VM actually restart such VM. This requires qubesd to handle the restarting all the relevant daemons (qrexec, qubesdb etc). That's also the time to reload VM configuration (maxmem, color etc). Due to that, libvirt should likely remain configured to `<on_reboot>destroy</on_reboot>`, but qubesd should react on shutdown reason in the libvirt event.

This may be opt-in feature (enabled with qvm-features or qvm-prefs).

### The value to a user, and who that user might be

Currently when executing "reboot" within a VM, it is actually shut down. This is especially annoying for Windows VMs which do perform reboot from time to time (some updates, etc).

**Comment by ben-grande** [comment]
Above [this line in qubes.app.Qubes._domain_event_callback()](https://github.com/QubesOS/qubes-core-admin/blob/71b5ab213ce1ca14af103a3c9f00a12b677c8113/qubes/app.py#L1616), added:

```python3
vm.log.info("LIBVIRT event: {}: {}: {}".format(event, _detail, _opaque)
```
Then restart qubesd.

For context, the [event types](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainEventType):

```c
enum virDomainEventType {
VIR_DOMAIN_EVENT_DEFINED	=	0 (0x0)	
VIR_DOMAIN_EVENT_UNDEFINED	=	1 (0x1)	
VIR_DOMAIN_EVENT_STARTED	=	2 (0x2)	
VIR_DOMAIN_EVENT_SUSPENDED	=	3 (0x3)	
VIR_DOMAIN_EVENT_RESUMED	=	4 (0x4)	
VIR_DOMAIN_EVENT_STOPPED	=	5 (0x5)	
VIR_DOMAIN_EVENT_SHUTDOWN	=	6 (0x6)	
VIR_DOMAIN_EVENT_PMSUSPENDED	=	7 (0x7)	
VIR_DOMAIN_EVENT_CRASHED	=	8 (0x8)	
VIR_DOMAIN_EVENT_LAST	=	9 (0x9)	
}
```

```c
enum virDomainEventStoppedDetailType {
VIR_DOMAIN_EVENT_STOPPED_SHUTDOWN	=	0 (0x0)	// Normal shutdown
VIR_DOMAIN_EVENT_STOPPED_DESTROYED	=	1 (0x1) // Forced poweroff from host
VIR_DOMAIN_EVENT_STOPPED_CRASHED	=	2 (0x2)	// Guest crashed
VIR_DOMAIN_EVENT_STOPPED_MIGRATED	=	3 (0x3)	// Migrated off to another host
VIR_DOMAIN_EVENT_STOPPED_SAVED	=	4 (0x4)	// Saved to a state file
VIR_DOMAIN_EVENT_STOPPED_FAILED	=	5 (0x5)	// Host emulator/mgmt failed
VIR_DOMAIN_EVENT_STOPPED_FROM_SNAPSHOT	=	6 (0x6)	// offline snapshot loaded
VIR_DOMAIN_EVENT_STOPPED_RECREATED	=	7 (0x7)	// Secure guest recreation
VIR_DOMAIN_EVENT_STOPPED_LAST	=	8 (0x8)	
}
```

---

Attempted to reboot the qube with `qvm-run -p -uroot test reboot now`:
```
INFO: vm.test: LIBVIRT event: 5: 0: None
```
It just shutdown, as expected, as libvirt `on_reboot` is set to `destroy`.

---

Modified `/usr/share/qubes/templates/libvirt/xen.xml`, changed `on_reboot` value from `destroy` to `restart`, and shutdown and started the domain to get new libvirt files. Then issued `qvm-run -p -uroot test reboot now`, exited 129 and qubesd logged:
```
INFO: vm.test: LIBVIRT event: 2: 0: None
INFO: vm.test: LIBVIRT event: 5: 0: None
WARNING: vm.test: Stopped event from libvirt received, but domain is in state Transient!
```
The domain did not shutdown. `virsh -c xen cpu-stats test` reports increasing CPU time, the `xl list` reports the domain with state `-b-----`. Attempting `qvm-shutdown test` succeeded and qubesd logged:
```
INFO: vm.test: LIBVIRT event: 5: 0: None
```
Hm... ok, I added `"Transient"` to the acceptable list, restarted qubesd, started the domain and issued the reboot of the domain and:

```
INFO: vm.test: LIBVIRT event: 2: 0: None
INFO: vm.test: LIBVIRT event: 5: 0: None
INFO: vm.test: LIBVIRT event: 5: 2: None
```
The domain only did shutdown. But we got a [detail](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainEventStoppedDetailType) this time, `VIR_DOMAIN_EVENT_STOPPED_CRASHED`. The domain did shutdown this time, still no reboot (obviously).

---

For another day: Register reboot event [VIR_DOMAIN_EVENT_ID_REBOOT](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainEventID) and see if it logs something without changing and with changes to `on_reboot` setting. Depending on the result, the next step might not be necessary. Next step: save `VIR_DOMAIN_EVENT_STARTED` to a variable when the qube state is `"Running"` (which is only fired once per guest boot), and use that to know that an imminent `VIR_DOMAIN_EVENT_STOPPED` will mean guest wants reboot of the domain.

On `QubesVM.start()`, it creates the domain, see `self.libvirt_domain.createWithFlags`, `qrexec` and `qubesdb` daemons starting, which means that the `start()` method is enough to restart them.

**Comment by marmarek** [comment]
One thing is detecting when reboot got requested, this I think you'll manage based on the above.

Then, full shutdown need to be handled (including `vm.storage.stop()` etc), and then normal startup. Basically what you described at the end.

As a nice optimization, memory allocation could be retained, to avoid situation when memory of a qube on shutdown gets redistributed to others, and then there is not enough to start it again. I imagine this could be done by telling qmemman about requested reboot, so it keeps that memory reserved, and then some modified `vm.request_mem()` could tell qmemman to take memory from that preserved allocation first.

**Comment by ben-grande** [comment]
I got it to reboot nicely. Only thing missing is qmemman and finishing the property definition.

> This may be opt-in feature (enabled with qvm-features or qvm-prefs).

- Just a global property or
- per-qube or
- per-qube with default to global (or default to specific template hierarchy if we want the behavior per OS)

**Comment by ben-grande** [comment]
> As a nice optimization, memory allocation could be retained, to avoid situation when memory of a qube on shutdown gets redistributed to others, and then there is not enough to start it again. I imagine this could be done by telling qmemman about requested reboot, so it keeps that memory reserved, and then some modified `vm.request_mem()` could tell qmemman to take memory from that preserved allocation first.

- I'm thinking of saving the xid in dom0 on reboot request prior to reboot (qmemman uses xid (domid), but it changes on reboot)
- Asking qmemman to not redistribute from that xid until timeout defined by qrexec_timeout
- Then send the old xid on `request_mem` on `QubesVM.start()`

I hope this is doable without too much changes on QMemMan, else, #1587 might be required.

---

Edit:

> * Asking qmemman to not redistribute from that xid

`del_domain` removes the domain and `/local/domain` only list running domains. I might need to save `/name`.

**Comment by marmarek** [comment]
> This may be opt-in feature (enabled with qvm-features or qvm-prefs).

Thinking about it more, maybe opt-in isn't even necessary, just have it supported by default.

**Comment by marmarek** [comment]
BTW, worth testing also on some non-Linux system in a HVM.

**Comment by ben-grande** [comment]
> BTW, worth testing also on some non-Linux system in a HVM.

On a Windows 11 HVM with QWT, when attempting to shutdown, it works, but when attempting to restart, raises exception, independently of using GUI or `shutdown /r /t 0`

```python3
May 19 17:35:42 dom0 qubesd[1886]: INFO: vm.windows-11: LIBVIRT event: 2: 0: None
May 19 17:35:42 dom0 qubesd[1886]: INFO: vm.windows-11: LIBVIRT event: 5: 0: None
May 19 17:35:50 dom0 qubesd[1886]: ERROR: vm.windows-11: Failed to stop storage
May 19 17:35:50 dom0 qubesd[1886]: Traceback (most recent call last):
May 19 17:35:50 dom0 qubesd[1886]:   File "/usr/lib/python3.13/site-packages/qubes/vm/qubesvm.py", line 1658, in on_domain_stopped
May 19 17:35:50 dom0 qubesd[1886]:     await self.storage.stop()
May 19 17:35:50 dom0 qubesd[1886]:   File "/usr/lib/python3.13/site-packages/qubes/storage/__init__.py", line 861, in stop
May 19 17:35:50 dom0 qubesd[1886]:     await qubes.utils.void_coros_maybe(
May 19 17:35:50 dom0 qubesd[1886]:     ...<9 lines>...
May 19 17:35:50 dom0 qubesd[1886]:     )
May 19 17:35:50 dom0 qubesd[1886]:   File "/usr/lib/python3.13/site-packages/qubes/utils.py", line 326, in void_coros_maybe
May 19 17:35:50 dom0 qubesd[1886]:     task.result()  # re-raises exception if task failed
May 19 17:35:50 dom0 qubesd[1886]:     ~~~~~~~~~~~^^
May 19 17:35:50 dom0 qubesd[1886]:   File "/usr/lib/python3.13/site-packages/qubes/storage/__init__.py", line 283, in wrapper
May 19 17:35:50 dom0 qubesd[1886]:     return await method(self, *args, **kwargs)
May 19 17:35:50 dom0 qubesd[1886]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
May 19 17:35:50 dom0 qubesd[1886]:   File "/usr/lib/python3.13/site-packages/qubes/storage/lvm.py", line 808, in stop
May 19 17:35:50 dom0 qubesd[1886]:     changed = await self._remove_if_exists(self.vid)
May 19 17:35:50 dom0 qubesd[1886]:               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
May 19 17:35:50 dom0 qubesd[1886]:   File "/usr/lib/python3.13/site-packages/qubes/storage/lvm.py", line 776, in _remove_if_exists
May 19 17:35:50 dom0 qubesd[1886]:     await qubes_lvm_coro(cmd, self.log)
May 19 17:35:50 dom0 qubesd[1886]:   File "/usr/lib/python3.13/site-packages/qubes/storage/lvm.py", line 979, in qubes_lvm_coro
May 19 17:35:50 dom0 qubesd[1886]:     return _process_lvm_output(p.returncode, out, err, log)
May 19 17:35:50 dom0 qubesd[1886]:   File "/usr/lib/python3.13/site-packages/qubes/storage/lvm.py", line 949, in _process_lvm_output
May 19 17:35:50 dom0 qubesd[1886]:     raise qubes.exc.StoragePoolException(err)
May 19 17:35:50 dom0 qubesd[1886]: qubes.exc.StoragePoolException:   Logical volume qubes_dom0/vm-windows-11-volatile in use.
May 19 17:35:50 dom0 qubesd[1886]: INFO: vm.windows-11: STARTING
May 19 17:35:50 dom0 qubesd[1886]: INFO: vm.windows-11: STARTING: inside lock
```

As `domain-stopped` just logs the exception and doesn't raise anything, `start()` is still called, not good, but not end of the world, it seems to return early, as the domain state is `Transient`. Still, I might make it an `exception`.

I logged `get_power_state()` and `libvirt_domain.state()` before `domain-stopped`

- For the shutdown case, it shows `Halted [5, 1]`, but it never enters the sleep loop, because the state is already `"Halted"` without having to wait sleep, so all good.
- For the shutdown case, it shows `Transient [1, 1]`

Then, inside the `sleep` loop, modified the `== "Dying"` to `!= "Transient".

- For restart, it shows `Transient [1, 1]`, eternally (longer than two minutes at least, while normal shutdown takes less than ten seconds), not good. When I run `qvm-kill windows-11`, it exits the loop and the vm is killed and started. Well, I will need to remove the ability to restart in case the qube was requested to be killed.

The above is true independent of qube, also fails on Linux non-hvm qube, eternally transient. Interesting. What I ended up with, is `libvirt_domain.destroy()` if `reboot and self.virt_mode == "hvm"`. The problem is not being HVM, because rebooting `sys-usb`, which is HVM, doesn't require such `destroy()`. Therefore, I am downloading OpenBSD 7.9, Ubuntu Desktop 26.04 and Windows 11 (without QWT this time), and will see if the problem is OS specific or not.

---

Downloaded OpenBSD and Ubuntu Desktop 26.04 and converted them into HVM standalones and finished their installation. When attempting to reboot, `QubesVM.on_libvirt_domain_stopped` returns early on both, and says that domain state is `Running [1, 1]`.... hmmm. I used this on OpenBSD: `reboot -q` (stands for quick), and used the GUI on Ubuntu tray, that said it could take 60 seconds.

Then, I gave up on the tray for Ubuntu, and tried to `sudo reboot now`, but gnome-session was inhibiting it. Then I used `sudo systemctl reboot -i` to ignore inhibitors, as the error message recommended, and it attempted to reboot, but same error as in windows, of volatile volume in use.

The only thing that worked for all non standard HVMs (Windows, Ubuntu, OpenBSD), was to call `await asyncio.to_thread(self.libvirt_domain.destroy)` on them, inside `_domain_stopped_coro`, before "domain-stopped".

---

I tried logging `VIR_DOMAIN_EVENT_ID_REBOOT`, but didn't receive anything under any circumstance or OS.

**Comment by marmarek** [comment]
Hm, with a change `<on_reboot>restart</on_reboot>`, you don't have a time window to do `storage.stop()` and `storage.start()`, as the VM gets restarted by libvirt without waiting to qubesd to do its thing. IIUC this also means the VM gets started not paused, right? This means qubesdb is not populated before starting the VM, only shortly after. In practice it will likely be early enough, but in theory it may happen that VM will wake up with empty qubesdb at boot, which will be quite bad.

Ideally, it would be better if libvirt would not restart VM on its own, but pass the shutdown reason in the event and let qubesd restart it. Maybe `preserve` action would allow that? (in that case, normal shutdown will need to call destroy manually)

**Comment by ben-grande** [comment]
For [domain format](https://libvirt.org/formatdomain.html) says:

> QEMU/KVM/HVF supports the on_poweroff and on_reboot events handling the destroy and restart actions ...

Doesn't mention Xen, so it might not be fully supported or just lacking documentation.

---

> Hm, with a change `<on_reboot>restart</on_reboot>`, you don't have a time window to do `storage.stop()` and `storage.start()`, as the VM gets restarted by libvirt without waiting to qubesd to do its thing. IIUC this also means the VM gets started not paused, right? This means qubesdb is not populated before starting the VM, only shortly after. In practice it will likely be early enough, but in theory it may happen that VM will wake up with empty qubesdb at boot, which will be quite bad.

I don't think libvirt is being able to start the qube again, it my PR code calling `self.start()` after all the shutdown events.

---

With `on_reboot` with `restart`, emits the `VIR_DOMAIN_STARTED` indicating that it wants reboot, works for standard qubes, while for HVMs requires `libvirt_domain.destroy()`:

<details><summary>Windows</summary>
<p>

```
May 20 16:43:40 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_event_callback: 2: 0: None
May 20 16:43:40 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_event_callback: 5: 0: None
May 20 16:43:40 dom0 qubesd[271684]: INFO: vm.windows-11: AAA on_libvirt_domain_stopped: Transient: [1, 1]
May 20 16:43:40 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_stopped_coro: reboot=True
May 20 16:43:40 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_stopped_coro: inside lock
May 20 16:43:40 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_stopped_coro: Transient: [1, 1]

# state changes as the script calls self.libvirt_domain.destroy() on certain HVMs. Without this, would fail to remove volatile volume
# It shows "Transient" because this domain has QWT, but Qrexec is not running anymore, still, the state is VIR_DOMAIN_RUNNING, VIR_DOMAIN_RUNNING_BOOTED

May 20 16:43:40 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_event_callback: 5: 1: None
May 20 16:43:40 dom0 qubesd[271684]: INFO: vm.windows-11: AAA on_libvirt_domain_stopped: Halted: [5, 2]

# 5,2 seems normal here: VIR_DOMAIN_SHUTOFF_DESTROYED

May 20 16:43:44 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_stopped_coro: finished with: Halted: [5, 2]
May 20 16:43:44 dom0 qubesd[271684]: INFO: vm.windows-11: AAA start: Halted: [5, 2]
May 20 16:43:44 dom0 qubesd[271684]: INFO: vm.windows-11: AAA start inside lock: Halted: [5, 2]
May 20 16:43:44 dom0 qubesd[271684]: INFO: vm.windows-11: AAA start after shutdown handled: Halted: [5, 2]
May 20 16:43:44 dom0 qubesd[271684]: INFO: vm.windows-11: Starting qube windows-11
May 20 16:43:46 dom0 qubesd[271684]: INFO: vm.windows-11: Setting Qubes DB info for the qube
May 20 16:43:46 dom0 qubesd[271684]: INFO: vm.windows-11: Starting Qubes DB
May 20 16:43:46 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_event_callback: 0: 1: None
May 20 16:43:46 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_event_callback: 2: 0: None
May 20 16:43:46 dom0 runuser[274475]: pam_unix(runuser:session): session opened for user user(uid=1000) by (uid=0)
May 20 16:43:46 dom0 runuser[274475]: pam_unix(runuser:session): session closed for user user
May 20 16:43:46 dom0 qubesd[271684]: INFO: vm.windows-11: Activating qube
May 20 16:43:46 dom0 qubesd[271684]: INFO: vm.windows-11: AAA _domain_event_callback: 4: 0: None
May 20 16:43:46 dom0 runuser[274481]: pam_unix(runuser:session): session opened for user user(uid=1000) by (uid=0)
May 20 16:43:46 dom0 runuser[274481]: pam_unix(runuser:session): session closed for user user
May 20 16:43:46 dom0 runuser[274495]: pam_unix(runuser:session): session opened for user user(uid=1000) by (uid=0)
May 20 16:44:06 dom0 runuser[274495]: pam_unix(runuser:session): session closed for user user
```

</p>
</details> 

<details><summary>Ubuntu</summary>
<p>

```
May 20 16:10:32 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_event_callback: 2: 0: None
May 20 16:10:32 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_event_callback: 5: 0: None
May 20 16:10:32 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA on_libvirt_domain_stopped: Running: [1, 1]
May 20 16:10:32 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: reboot=True
May 20 16:10:32 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: inside lock
May 20 16:10:32 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: Running: [1, 1]

# state changes as the script calls self.libvirt_domain.destroy() on certain HVMs. Without this, would fail to remove volatile volume
# It shows "Running" as "on_domain_is_fully_usable" returns True, as this qube doesn't have Qrexec, still, the state is VIR_DOMAIN_RUNNING, VIR_DOMAIN_RUNNING_BOOTED

May 20 16:10:33 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_event_callback: 5: 1: None
May 20 16:10:33 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA on_libvirt_domain_stopped: Halted: [5, 2]

# 5,2 seems normal here: VIR_DOMAIN_SHUTOFF_DESTROYED

May 20 16:10:36 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: finished with: Halted: [5, 2]
May 20 16:10:36 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA start: Halted: [5, 2]
May 20 16:10:36 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA start inside lock: Halted: [5, 2]
May 20 16:10:36 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA start after shutdown handled: Halted: [5, 2]
May 20 16:10:36 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: Starting qube ubuntu-26.04
May 20 16:10:38 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: Setting Qubes DB info for the qube
May 20 16:10:38 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: Starting Qubes DB
May 20 16:10:38 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_event_callback: 0: 1: None
May 20 16:10:38 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_event_callback: 2: 0: None
May 20 16:10:38 dom0 runuser[220218]: pam_unix(runuser:session): session opened for user user(uid=1000) by (uid=0)
May 20 16:10:38 dom0 runuser[220218]: pam_unix(runuser:session): session closed for user user
May 20 16:10:38 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: Activating qube
May 20 16:10:38 dom0 qubesd[213791]: INFO: vm.ubuntu-26.04: AAA _domain_event_callback: 4: 0: None
May 20 16:10:38 dom0 runuser[220221]: pam_unix(runuser:session): session opened for user user(uid=1000) by (uid=0)
May 20 16:10:38 dom0 runuser[220221]: pam_unix(runuser:session): session closed for user user
```

</p>
</details>

<details><summary>sys-firewall</summary>
<p>

```
May 20 16:16:35 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA start: Running: [1, 5]
May 20 16:16:35 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA start inside lock: Running: [1, 5]
May 20 16:16:48 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_event_callback: 2: 0: None
May 20 16:16:48 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_event_callback: 5: 0: None
May 20 16:16:48 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA on_libvirt_domain_stopped: Transient: [1, 1]
May 20 16:16:48 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: reboot=True
May 20 16:16:48 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: inside lock
May 20 16:16:48 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: Transient: [1, 1]

# It shows "Transient", with state VIR_DOMAIN_RUNNING, VIR_DOMAIN_RUNNING_BOOTED, but the qube is not using the volumes anymore as domain-stopped will be called without errors.

May 20 16:16:49 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_event_callback: 5: 2: None
May 20 16:16:49 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA on_libvirt_domain_stopped: Halted: [5, 3]

# 5,3 seems strange here: VIR_DOMAIN_SHUTOFF_CRASHED, but the restart works.

May 20 16:16:50 dom0 qubesd[213791]: WARNING: vm.disp269: Deferred detaching libvirt net device because qube is paused
May 20 16:16:50 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: finished with: Halted: [5, 3]
May 20 16:16:50 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA start: Halted: [5, 3]
May 20 16:16:50 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA start inside lock: Halted: [5, 3]
May 20 16:16:50 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA start after shutdown handled: Halted: [5, 3]
May 20 16:16:50 dom0 qubesd[213791]: INFO: vm.sys-firewall: Starting qube sys-firewall
May 20 16:16:51 dom0 qubesd[213791]: INFO: vm.sys-firewall: Setting Qubes DB info for the qube
May 20 16:16:51 dom0 qubesd[213791]: INFO: vm.sys-firewall: Starting Qubes DB
May 20 16:16:51 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_event_callback: 0: 1: None
May 20 16:16:51 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_event_callback: 2: 0: None
May 20 16:16:51 dom0 runuser[229717]: pam_unix(runuser:session): session opened for user user(uid=1000) by (uid=0)
May 20 16:16:51 dom0 runuser[229717]: pam_unix(runuser:session): session closed for user user
May 20 16:16:51 dom0 qubesd[213791]: INFO: vm.sys-firewall: Activating qube
May 20 16:16:51 dom0 qubesd[213791]: INFO: vm.sys-firewall: AAA _domain_event_callback: 4: 0: None
```

</p>
</details>

---

Using `on_reboot` with `preserve`, none emit an event saying it wants to reboot:

<details><summary>Windows</summary>
<p>

```
May 20 16:36:14 dom0 qubesd[249760]: INFO: vm.windows-11: AAA _domain_event_callback: 5: 0: None
May 20 16:36:14 dom0 qubesd[249760]: INFO: vm.windows-11: AAA on_libvirt_domain_stopped: Dying: [5, 1]
May 20 16:36:14 dom0 qubesd[249760]: INFO: vm.windows-11: AAA _domain_stopped_coro: reboot=False
May 20 16:36:14 dom0 qubesd[249760]: INFO: vm.windows-11: AAA _domain_stopped_coro: inside lock
May 20 16:36:14 dom0 qubesd[249760]: INFO: vm.windows-11: AAA _domain_stopped_coro: looping: Dying: [5, 1]
# multiple loops
May 20 16:36:23 dom0 qubesd[249760]: INFO: vm.windows-11: AAA _domain_stopped_coro: looping: Dying: [5, 1]

# I called qvm-kill sys-firewall to quit the loop

May 20 16:36:24 dom0 qubesd[249760]: INFO: vm.windows-11: AAA _domain_stopped_coro: Halted: [5, 2]
May 20 16:36:24 dom0 qubesd[249760]: INFO: vm.windows-11: AAA _domain_event_callback: 5: 1: None
May 20 16:36:24 dom0 qubesd[249760]: INFO: vm.windows-11: AAA on_libvirt_domain_stopped: Halted: [5, 2]
May 20 16:36:25 dom0 qubesd[249760]: INFO: vm.windows-11: AAA _domain_stopped_coro: finished with: Halted: [5, 2]
```

</p>
</details>

<details><summary>Ubuntu</summary>
<p>

```
May 20 16:39:01 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA _domain_event_callback: 5: 0: None
May 20 16:39:01 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA on_libvirt_domain_stopped: Dying: [5, 1]
May 20 16:39:01 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: reboot=False
May 20 16:39:01 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: inside lock
May 20 16:39:01 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: looping: Dying: [5, 1]
# mutiple loops
May 20 16:39:16 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: looping: Dying: [5, 1]

# I called qvm-kill sys-firewall to quit the loop

May 20 16:39:16 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: Halted: [5, 2]
May 20 16:39:16 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA _domain_event_callback: 5: 1: None
May 20 16:39:16 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA on_libvirt_domain_stopped: Halted: [5, 2]
May 20 16:39:17 dom0 qubesd[249760]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: finished with: Halted: [5, 2]
```

</p>
</details>

<details><summary>sys-firewall</summary>
<p>

```
May 20 16:29:28 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA _domain_event_callback: 5: 0: None
May 20 16:29:28 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA on_libvirt_domain_stopped: Dying: [5, 1]
May 20 16:29:28 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: reboot=False
May 20 16:29:28 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: inside lock
May 20 16:29:28 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: looping: Dying: [5, 1]
# multiple loops
May 20 16:29:34 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: looping: Dying: [5, 1]

# I called qvm-kill sys-firewall to quit the loop

May 20 16:29:34 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: Halted: [5, 2]
May 20 16:29:34 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA _domain_event_callback: 5: 1: None
May 20 16:29:34 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA on_libvirt_domain_stopped: Halted: [5, 2]
May 20 16:29:34 dom0 qubesd[249760]: WARNING: vm.disp269: Deferred detaching libvirt net device because qube is paused
May 20 16:29:34 dom0 qubesd[249760]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: finished with: Halted: [5, 2]
```

</p>
</details> 

---

Using `on_reboot` with `destroy`, none emit an event saying it wants to reboot:

<details><summary>Windows</summary>
<p>

```
May 20 17:10:00 dom0 qubesd[304009]: INFO: vm.windows-11: AAA _domain_event_callback: 5: 0: None
May 20 17:10:00 dom0 qubesd[304009]: INFO: vm.windows-11: AAA on_libvirt_domain_stopped: Halted: [5, 1]
May 20 17:10:00 dom0 qubesd[304009]: INFO: vm.windows-11: AAA _domain_stopped_coro: reboot=False
May 20 17:10:00 dom0 qubesd[304009]: INFO: vm.windows-11: AAA _domain_stopped_coro: inside lock
May 20 17:10:00 dom0 qubesd[304009]: INFO: vm.windows-11: AAA _domain_stopped_coro: Halted: [5, 1]

# State is "Halted" already, not "Dying" eternally like in "<on_reboot>preserve</on_reboot>."

May 20 17:10:04 dom0 qubesd[304009]: INFO: vm.windows-11: AAA _domain_stopped_coro: finished with: Halted: [5, 1]
```

</p>
</details> 

<details><summary>Ubuntu</summary>
<p>

```
May 20 17:08:22 dom0 qubesd[304009]: INFO: vm.ubuntu-26.04: AAA _domain_event_callback: 5: 0: None
May 20 17:08:22 dom0 qubesd[304009]: INFO: vm.ubuntu-26.04: AAA on_libvirt_domain_stopped: Halted: [5, 1]
May 20 17:08:22 dom0 qubesd[304009]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: reboot=False
May 20 17:08:22 dom0 qubesd[304009]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: inside lock
May 20 17:08:22 dom0 qubesd[304009]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: Halted: [5, 1]

# State is "Halted" already, not "Dying" eternally like in "<on_reboot>preserve</on_reboot>."

May 20 17:08:23 dom0 qubesd[304009]: INFO: vm.ubuntu-26.04: AAA _domain_stopped_coro: finished with: Halted: [5, 1]
```

</p>
</details> 

<details><summary>sys-firewall</summary>
<p>

```
May 20 17:05:03 dom0 qubesd[304009]: INFO: vm.sys-firewall: AAA _domain_event_callback: 5: 0: None
May 20 17:05:03 dom0 qubesd[304009]: INFO: vm.sys-firewall: AAA on_libvirt_domain_stopped: Halted: [5, 1]
May 20 17:05:03 dom0 qubesd[304009]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: reboot=False
May 20 17:05:03 dom0 qubesd[304009]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: inside lock
May 20 17:05:03 dom0 qubesd[304009]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: Halted: [5, 1]

# State is "Halted" already, not "Dying" eternally like in "<on_reboot>preserve</on_reboot>."

May 20 17:05:04 dom0 qubesd[304009]: WARNING: vm.disp269: Deferred detaching libvirt net device because qube is paused
May 20 17:05:04 dom0 qubesd[304009]: INFO: vm.sys-firewall: AAA _domain_stopped_coro: finished with: Halted: [5, 1]
```

</p>
</details>

**Comment by marmarek** [comment]
I'm not sure, but I think with on_reboot=restart it tries to restart it on its own and possibly fails...

**Comment by rustybird** [comment]
> Thinking about it more, maybe opt-in isn't even necessary, just have it supported by default.

One bad thing a VM would be able to do with this is throw away all previous revisions of its storage volumes, simply by restarting itself a couple of times. So a user could no longer rely on being able to revert in case of compromise (or an unintentional restart loop). :(

**Comment by ben-grande** [comment]
It's not related to HVMs at all. Volume removal of `domain-stopped` event always fails  on Standalones and Templates with `Transient: [1, 1]` at that stage, even if adding a long sleep before `domain-stopped` being fired, only thing that solved was `libvirt_domain.destroy()`.

<details><summary>app qube</summary>
<p>

```
May 20 18:11:49 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_event_callback: 2: 0: None: Transient: [1, 1]
May 20 18:11:49 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_event_callback: 5: 0: None: Transient: [1, 1]
May 20 18:11:49 dom0 qubesd[422368]: INFO: vm.work: AAA on_libvirt_domain_stopped: Transient: [1, 1]
May 20 18:11:49 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_stopped_coro: reboot=True
May 20 18:11:49 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_stopped_coro: inside lock
May 20 18:11:49 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_stopped_coro prior domain-stopped: Transient: [1, 1]
May 20 18:11:49 dom0 qubesd[422368]: INFO: vm.work: AAA domain-stopped: Transient: [1, 1]
May 20 18:11:51 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_event_callback: 5: 2: None: Halted: [5, 3]

# Notice the event above is received only after we are inside `domain-stopped`.

May 20 18:11:51 dom0 qubesd[422368]: INFO: vm.work: AAA on_libvirt_domain_stopped: Halted: [5, 3]
May 20 18:11:52 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_stopped_coro after domain-stopped: Halted: [5, 3]
May 20 18:11:52 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_stopped_coro: finished with: Halted: [5, 3]
May 20 18:11:52 dom0 qubesd[422368]: INFO: vm.work: AAA start: Halted: [5, 3]
May 20 18:11:52 dom0 qubesd[422368]: INFO: vm.work: AAA start inside lock: Halted: [5, 3]
May 20 18:11:52 dom0 qubesd[422368]: INFO: vm.work: AAA start after shutdown handled: Halted: [5, 3]
May 20 18:11:52 dom0 qubesd[422368]: INFO: vm.work: Starting qube work
May 20 18:11:53 dom0 qubesd[422368]: INFO: vm.work: Setting Qubes DB info for the qube
May 20 18:11:53 dom0 qubesd[422368]: INFO: vm.work: Starting Qubes DB
May 20 18:11:53 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_event_callback: 0: 1: None: Paused: [3, 1]
May 20 18:11:53 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_event_callback: 2: 0: None: Paused: [3, 1]
May 20 18:11:53 dom0 runuser[423702]: pam_unix(runuser:session): session opened for user user(uid=1000) by (uid=0)
May 20 18:11:53 dom0 runuser[423702]: pam_unix(runuser:session): session closed for user user
May 20 18:11:53 dom0 qubesd[422368]: INFO: vm.work: Activating qube
May 20 18:11:53 dom0 qubesd[422368]: INFO: vm.work: AAA _domain_event_callback: 4: 0: None: Transient: [1, 5]
May 20 18:11:53 dom0 runuser[423715]: pam_unix(runuser:session): session opened for user user(uid=1000) by (uid=0)
May 20 18:11:57 dom0 runuser[423715]: pam_unix(runuser:session): session closed for user user
```

</p>
</details> 

<details><summary>standalone</summary>
<p>

```
May 20 18:14:22 dom0 qubesd[422368]: INFO: vm.test-std: AAA _domain_event_callback: 2: 0: None: Transient: [1, 1]
May 20 18:14:22 dom0 qubesd[422368]: INFO: vm.test-std: AAA _domain_event_callback: 5: 0: None: Transient: [1, 1]
May 20 18:14:22 dom0 qubesd[422368]: INFO: vm.test-std: AAA on_libvirt_domain_stopped: Transient: [1, 1]
May 20 18:14:22 dom0 qubesd[422368]: INFO: vm.test-std: AAA _domain_stopped_coro: reboot=True
May 20 18:14:22 dom0 qubesd[422368]: INFO: vm.test-std: AAA _domain_stopped_coro: inside lock
May 20 18:14:22 dom0 qubesd[422368]: INFO: vm.test-std: AAA _domain_stopped_coro prior domain-stopped: Transient: [1, 1]
May 20 18:14:22 dom0 qubesd[422368]: INFO: vm.test-std: AAA domain-stopped: Transient: [1, 1]

# Notice it takes some time here, in this case, 8 seconds to failure. I event tried adding 30s sleep and it didn't help. Only thing that helped was libvirt_domain.destroy().

May 20 18:14:30 dom0 qubesd[422368]: ERROR: vm.test-std: Failed to stop storage
May 20 18:14:30 dom0 qubesd[422368]: Traceback (most recent call last):
May 20 18:14:30 dom0 qubesd[422368]:   File "/usr/lib/python3.13/site-packages/qubes/vm/qubesvm.py", line 1711, in on_domain_stopped
May 20 18:14:30 dom0 qubesd[422368]:     await self.storage.stop()
May 20 18:14:30 dom0 qubesd[422368]:   File "/usr/lib/python3.13/site-packages/qubes/storage/__init__.py", line 861, in stop
May 20 18:14:30 dom0 qubesd[422368]:     await qubes.utils.void_coros_maybe(
May 20 18:14:30 dom0 qubesd[422368]:     ...<9 lines>...
May 20 18:14:30 dom0 qubesd[422368]:     )
May 20 18:14:30 dom0 qubesd[422368]:   File "/usr/lib/python3.13/site-packages/qubes/utils.py", line 326, in void_coros_maybe
May 20 18:14:30 dom0 qubesd[422368]:     task.result()  # re-raises exception if task failed
May 20 18:14:30 dom0 qubesd[422368]:     ~~~~~~~~~~~^^
May 20 18:14:30 dom0 qubesd[422368]:   File "/usr/lib/python3.13/site-packages/qubes/storage/__init__.py", line 283, in wrapper
May 20 18:14:30 dom0 qubesd[422368]:     return await method(self, *args, **kwargs)
May 20 18:14:30 dom0 qubesd[422368]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
May 20 18:14:30 dom0 qubesd[422368]:   File "/usr/lib/python3.13/site-packages/qubes/storage/lvm.py", line 808, in stop
May 20 18:14:30 dom0 qubesd[422368]:     changed = await self._remove_if_exists(self.vid)
May 20 18:14:30 dom0 qubesd[422368]:               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
May 20 18:14:30 dom0 qubesd[422368]:   File "/usr/lib/python3.13/site-packages/qubes/storage/lvm.py", line 776, in _remove_if_exists
May 20 18:14:30 dom0 qubesd[422368]:     await qubes_lvm_coro(cmd, self.log)
May 20 18:14:30 dom0 qubesd[422368]:   File "/usr/lib/python3.13/site-packages/qubes/storage/lvm.py", line 979, in qubes_lvm_coro
May 20 18:14:30 dom0 qubesd[422368]:     return _process_lvm_output(p.returncode, out, err, log)
May 20 18:14:30 dom0 qubesd[422368]:   File "/usr/lib/python3.13/site-packages/qubes/storage/lvm.py", line 949, in _process_lvm_output
May 20 18:14:30 dom0 qubesd[422368]:     raise qubes.exc.StoragePoolException(err)
May 20 18:14:30 dom0 qubesd[422368]: qubes.exc.StoragePoolException:   Logical volume qubes_dom0/vm-test-std-volatile in use.
May 20 18:14:30 dom0 qubesd[422368]: INFO: vm.test-std: AAA _domain_stopped_coro after domain-stopped: Transient: [1, 1]
May 20 18:14:30 dom0 qubesd[422368]: INFO: vm.test-std: AAA _domain_stopped_coro: finished with: Transient: [1, 1]
May 20 18:14:30 dom0 qubesd[422368]: INFO: vm.test-std: AAA start: Transient: [1, 1]
May 20 18:14:30 dom0 qubesd[422368]: INFO: vm.test-std: AAA start inside lock: Transient: [1, 1]
```

</p>
</details> 

---

Now `qvm-shutdown`:

<details><summary>app qube</summary>
<p>

```
May 20 18:40:44 dom0 qubesd[471256]: INFO: vm.work-1: AAA _domain_event_callback: 5: 0: None: Halted: [5, 1]
May 20 18:40:44 dom0 qubesd[471256]: INFO: vm.work-1: AAA on_libvirt_domain_stopped: Halted: [5, 1]
May 20 18:40:44 dom0 qubesd[471256]: INFO: vm.work-1: AAA _domain_stopped_coro: reboot=False
May 20 18:40:44 dom0 qubesd[471256]: INFO: vm.work-1: AAA _domain_stopped_coro: inside lock
May 20 18:40:44 dom0 qubesd[471256]: INFO: vm.work-1: AAA _domain_stopped_coro prior domain-stopped: Halted: [5, 1]
May 20 18:40:44 dom0 qubesd[471256]: INFO: vm.work-1: AAA domain-stopped: Halted: [5, 1]
May 20 18:40:45 dom0 qubesd[471256]: INFO: vm.work-1: AAA _domain_stopped_coro after domain-stopped: Halted: [5, 1]
May 20 18:40:45 dom0 qubesd[471256]: INFO: vm.work-1: AAA _domain_stopped_coro: finished with: Halted: [5, 1]
```

</p>
</details> 

<details><summary>standalone</summary>
<p>

```
May 20 18:41:33 dom0 qubesd[471256]: INFO: vm.test-std: AAA _domain_event_callback: 5: 0: None: Halted: [5, 1]
May 20 18:41:33 dom0 qubesd[471256]: INFO: vm.test-std: AAA on_libvirt_domain_stopped: Halted: [5, 1]
May 20 18:41:33 dom0 qubesd[471256]: INFO: vm.test-std: AAA _domain_stopped_coro: reboot=False
May 20 18:41:33 dom0 qubesd[471256]: INFO: vm.test-std: AAA _domain_stopped_coro: inside lock
May 20 18:41:33 dom0 qubesd[471256]: INFO: vm.test-std: AAA _domain_stopped_coro prior domain-stopped: Halted: [5, 1]
May 20 18:41:33 dom0 qubesd[471256]: INFO: vm.test-std: AAA domain-stopped: Halted: [5, 1]
May 20 18:41:36 dom0 qubesd[471256]: INFO: vm.test-std: AAA _domain_stopped_coro after domain-stopped: Halted: [5, 1]
May 20 18:41:36 dom0 qubesd[471256]: INFO: vm.test-std: AAA _domain_stopped_coro: finished with: Halted: [5, 1]
```

</p>
</details> 

---

Hm. I don't like having to call `.destroy`, but maybe it's safe (for the domain) at that stage.

---

> One bad thing a VM would be able to do with this is throw away all previous revisions of its storage volumes, simply by restarting itself a couple of times. So a user could no longer rely on being able to revert in case of compromise (or an unintentional restart loop). :(

True... unless I add yet another property of limiting the amount of reboots a qube can try, but unfortunately, not simple to limit it to `x < revisions_to_keep`, because Windows installation requires a lot of reboots, and at that time, `qvm-features` is not populated as QWT is not installed and therefore `os` feature is not set.

**Comment by ben-grande** [comment]
https://lists.libvirt.org/archives/list/devel@lists.libvirt.org/thread/ZB5ALEFMTONTYZMJ6R2H6JWTOFL4ETER/

**Comment by marmarek** [comment]
> True... unless I add yet another property of limiting the amount of reboots a qube can try, but unfortunately, not simple to limit it to `x < revisions_to_keep`, because Windows installation requires a lot of reboots, and at that time, `qvm-features` is not populated as QWT is not installed and therefore `os` feature is not set.

A rate limit would be useful indeed. Maybe some time-based, so user has a chance to stop it? Like, at most one reboot in a minute, or 5min?

BTW, ability for a VM to reboot would be useful for SELinux relabeling too. When full relabel happens, startup scripts trigger reboot after that. Right now it looks to the user as a failed startup...

**Comment by crass** [comment]
I was recently bitten by this quirk.

I'm not familiar with how revision counting is done, but it seems a solution to the issue of revisions being blown away would be to introduce simple revision segmentation. If revision counting is done by counting the number of volumes matching a specific filename format (ie `volume.timestamp`), then these in-VM generated revisions could have a non-matching format (eg. `volume.restart.timestamp`). They could then be cleaned up on the generation of the next regular user-initiated revision.

I believe it could be desirable in some circumstances to keep a revision for each reboot. Probably not generally useful though. It might be deemed reasonable to not have revisions be generated in this circumstance as an alternative solution to the above issue.

Also, I haven't tried this, but I would expect in Linux this could be mitigated within the VM by doing a kexec reboot instead of a firmware reboot. Would it be desirable to have reboot by default do a kexec reboot in the templates?

**Comment by ben-grande** [comment]
Back at working on this issue.

> I'm not familiar with how revision counting is done, but it seems a solution to the issue of revisions being blown away would be to introduce simple revision segmentation. If revision counting is done by counting the number of volumes matching a specific filename format (ie `volume.timestamp`), then these in-VM generated revisions could have a non-matching format (eg. `volume.restart.timestamp`). They could then be cleaned up on the generation of the next regular user-initiated revision.

We can't distinguish a "user-initiated revision", a shutdown is a shutdown no matter how it was created. We can distinguish reboot from shutdown (once I send a patch to libvirt), but using a separate revision tracking would miss integration in the admin API for restoring those volumes, would also require adapting the client tools to support restoration from those. I prefer to continue using the current revision system, with a time-based approach suggested by Marek.

> I believe it could be desirable in some circumstances to keep a revision for each reboot. Probably not generally useful though. It might be deemed reasonable to not have revisions be generated in this circumstance as an alternative solution to the above issue.

We won't disable revisions.

> Also, I haven't tried this, but I would expect in Linux this could be mitigated within the VM by doing a kexec reboot instead of a firmware reboot. Would it be desirable to have reboot by default do a kexec reboot in the templates?

This is a separate feature request, to enable kexec reboot. Also, I don't think it notifies the hypervisor that the qube is rebooting, meaning that triggers for `domain-shutdown` and `domain-start` won't run (doesn't disconnect and reconnect to netvm), as well as any thing that would normally be changed in a qube after a shutdown+start, such as volumes. A full restart is necessary for #8070 (deferred templates) for example.

**Comment by crass** [comment]
> Back at working on this issue.
> 
> > I'm not familiar with how revision counting is done, but it seems a solution to the issue of revisions being blown away would be to introduce simple revision segmentation. If revision counting is done by counting the number of volumes matching a specific filename format (ie `volume.timestamp`), then these in-VM generated revisions could have a non-matching format (eg. `volume.restart.timestamp`). They could then be cleaned up on the generation of the next regular user-initiated revision.
> 
> We can't distinguish a "user-initiated revision", a shutdown is a shutdown no matter how it was created.

"Can't" or "don't". The later maybe true, the former is highly questionable. As stated, that statement appears false to me. Can a user initiated shutdown be differentiated from a user initiated kill, from a user initiated restart, from a VM crash? I believe the answer to all of these is yes.

Even if your statement were true, its not relevant. The relevant question is "Can a shutdown be distinguished from a reboot?" And here, I believe the answer is yes as well.

> We can distinguish reboot from shutdown (once I send a patch to libvirt), but using a separate revision tracking would miss integration in the admin API for restoring those volumes, would also require adapting the client tools to support restoration from those. 

You make this sound like a big deal. I haven't reviewed the code, but I have a hard time imagining it would be that big of a change. I've also not had much experience with the revision selection UX, but if it just shows a list of revisions to choose to restore from, what's so difficult with showing, for instance, all snapshots that have a name prefix of `revision`. Let the user decide? Of course there could be some help text describing generally what certain revision tags mean.

Fun fact: The ZFS pool driver has a `cause` as part of the revision name which seems fit for this purpose and using a similar naming scheme as I proposed.

> I prefer to continue using the current revision system, with a time-based approach suggested by Marek.

Marek's suggestion is not mutually exclusive to my proposal. Its not an either/or choice. He's talking about rate limiting and I'm suggesting classifying and handling specially certain classifications. And how does rate limiting solve the issue when `revisions_to_keep` is defaulted to `1`? I suppose this whole issue of revision handling should be a separate linked subissue.

> > I believe it could be desirable in some circumstances to keep a revision for each reboot. Probably not generally useful though. It might be deemed reasonable to not have revisions be generated in this circumstance as an alternative solution to the above issue.
> 
> We won't disable revisions.

This isn't about "disabling revisions", its about defining when revisions are created. Revisions are not "disabled" just because the policy doesn't define them to be created under certain conditions. If they policy is to create revisions on shutdown, but not every 5 minutes (or on reboot), they are not disabled.

Qubes currently has no "in-VM reboot" revision policy because "in-VM reboot" is broken. So it becomes a "shutdown" which Qubes does have a policy for, and changing that policy is not what I am discussing.

What should be clearly defined first is, "what is a reboot?" and then "does a revision get created on reboot?"

> > Also, I haven't tried this, but I would expect in Linux this could be mitigated within the VM by doing a kexec reboot instead of a firmware reboot. Would it be desirable to have reboot by default do a kexec reboot in the templates?
> 
> This is a separate feature request, to enable kexec reboot.

Actually, kexec reboot is more on topic for this issue than all the revision stuff above. The issue is `Make in-VM reboot request actually restart the VM` and kexec reboot is one way to do that, though not completely general.

> Also, I don't think it notifies the hypervisor that the qube is rebooting, meaning that triggers for `domain-shutdown` and `domain-start` won't run (doesn't disconnect and reconnect to netvm), as well as any thing that would normally be changed in a qube after a shutdown+start, such as volumes.

I wouldn't expect it would notify the hypervisor, and I wouldn't expect a reboot in Qemu to notify the hypervisor either.

Why does a disconnect/reconnect to a netvm need to happen? Can't the previous connection be reused? I'm not familiar with much of Xen internals/usage.

Also, why would a reboot be expected to be a `shutdown+start`? On physical hardware there's the concept of cold vs. warm boot (aka reboot). When a user "reboots" their computer it is a warm boot. This issue is about "in-VM" reboot, which would map to a warm reboot. A VM/Qube shutdown+start is a cold boot. If you reboot in a Qemu, the process does not die and get recreated. A systemd reboot can be many things, but I believe by default it does not involved a power-off.

I think defining how an "in-VM" **should** behave is in order. I think we can all agree that whatever that definition is/will be, currently its broken in Qubes.

> A full restart is necessary for [#8070](https://github.com/QubesOS/qubes-issues/issues/8070) (deferred templates) for example.

What is a "full restart"? A cold boot? If not, what kind of warm boot? Keep in mind that kexec is a type of warm boot.

**Comment by ben-grande** [comment]
> As a nice optimization, memory allocation could be retained, to avoid situation when memory of a qube on shutdown gets redistributed to others, and then there is not enough to start it again.

Do you want to maintain the full memory allocation? It will freeze `mem_actual` which might be 4G, or do you want to freeze up to the initial memory with overhead that is used in `QubesVM.request_mem`? Let's say other domains want to start between a qube rebooting, than reserving too much memory might prevent other domains from starting. If no domain wants to start in the meantime, and no other action is triggering redistributing memory, than reserving just the initial memory might overkill. I am inclined to use the initial memory with overhead instead of current allocation.

**Comment by marmarek** [comment]
> I am inclined to use the initial memory with overhead instead of current allocation.

That makes sense, especially since the qube quite likely will report low usage shortly after anyway (until some bigger application is started inside).
